src/plans/solver_state.jl

```diff
@@ -205,65 +225,128 @@ iteration, i.e. acts on `(p,o,i)`, where `p` is a [`AbstractManoptProblem`](@ref
 Initialize the Functor to an (empty) set of keys, where `once` determines
 whether more that one update per iteration are effective
 
-    AbstractStateAction(keys, once=true])
+    function StoreStateAction(
+        general_keys::Vector{Symbol}=Symbol[],
+        point_values::NamedTuple=NamedTuple(),
+        tangent_values::NamedTuple=NamedTuple(),
+        once::Bool=true,
+    ))
```

Comment:
I tried for about an hour to understand how I would actually call this, I am not able to understand this.
Can we maybe do a concrete example here how the `NamedTuples` are used? It also implies that we always have to know at-time-of-construction all fields we want to store in points/tuples? This can not grow?

The keys of the NamedTuples are the fields we do storage for (or `:Gradient`/`:Iterate`)? All this would be good to be documented here.

Reply:
I think this will be easier to explain in a video call again :wink: .

The fast storage can't grow in this design but we still can use `general_keys` if we need to add things dynamically. Moreover, we need to have a point/vector for initialization at-time-of-construction. I could make it slightly slower and growable, I could also make it slightly slower and only require type of a point at time of construction if that would be useful.

Reply:
I think now that I slowly understand it, there is mainly one concern I have (besides the constructor but I that I already mentioned below.

Currently we have a tradeoff: Either we have points/vectors and are fast or we do not have those and are slow. Is there maybe also a trick to resolve the rest?
Now, if we once decide we could do the same with - say - cotangent vectors we again have to write all the accessor functions for example.
I was first thinking of making a dictionary of named tuples? or a named tuple of named tuples? Ah both would not work for the “number-storage” when we just store floats or ints.

Reply:
> Currently we have a tradeoff: Either we have points/vectors and are fast or we do not have those and are slow. Is there maybe also a trick to resolve the rest?

Sure, we can have efficient storage for other things as well but I haven't noticed other things being used that much so I haven't worked on that.

> Now, if we once decide we could do the same with - say - cotangent vectors we again have to write all the accessor functions for example.

I can unify accessors by differentiating storage pools by a type wrapper, so instead of `get_point_storage(u.storage, :Iterate)` we would have something like `get_storage(u.storage, PointStorage(:Iterate))`.

> I was first thinking of making a dictionary of named tuples? or a named tuple of named tuples? Ah both would not work for the “number-storage” when we just store floats or ints.

Dictionary won't work because it couldn't always be type-stable, for example. Nested named tuples would work but the code would be quite convoluted (even more convoluted than it is now). They could store raw numbers if we modified the tuple itself instead of its content but that's another level of convolution.

Reply:
I like the storage thing – and sure if it is not possible more generally, then I am fine, if we keep the general accessors, like you just prosed; because then, _if_ we come up with an efficient way and can store, for example, cotangents also easily, we do not have to change the access/store function interfaces.